### PR TITLE
Autobahn Test Merge

### PR DESCRIPTION
**Pull request by felix-re**
Includes all files needed to execute the autobahn testsuite and test the websocket implementation

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -76,6 +76,7 @@ SET(CJET_FILES
         socket_peer.c
         table.c
         timer.c
+	utf8_checker.c
```
Comment (gatzka) [inline src/CMakeLists.txt]: Fix indentation by using spaces instead of tabs.

File: src/autobahnfiles/abMain.c [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+#include "table.h"
+
+int main(int argc, char **argv)
+{
+	struct cmdline_config config = {
+        .run_foreground = true,
+	    .bind_local_only = false,
+	    .user_name = NULL,
+	    .passwd_file = NULL,
+        .request_target = "/",  //"/api/jet/",
```
Comment (gatzka) [inline src/autobahnfiles/abMain.c]: Please fix indentation.

File: src/autobahnfiles/abMain.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+	int c;
+
+	while ((c = getopt(argc, argv, "flp:r:u:")) != -1) {
+		switch (c) {
+		case 'f':
+			config.run_foreground = true;
+			break;
+		case 'l':
+            config.bind_local_only = false;
```
Comment (gatzka) [inline src/autobahnfiles/abMain.c]: Please fix indentation.

File: src/autobahnfiles/abMain.c [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+		goto element_hashtable_create_failed;
+	}
+
+	struct eventloop_epoll eloop = {
+	    .epoll_fd = 0,
+	    .loop = {
+	        .this_ptr = &eloop,
+	        .init = eventloop_epoll_init,
+	        .destroy = eventloop_epoll_destroy,
+            .run = eventloop_epoll_run,
```
Comment (gatzka) [inline src/autobahnfiles/abMain.c]: Please fix indentation.

File: src/autobahnfiles/abWebsocket_peer.c [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
+#endif
+
+#ifndef MIN
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+#endif
+
+#define CRLF "\r\n"
+
+#define WS_CONTINUATION_FRAME 0x0
```
Comment (gatzka) [inline src/autobahnfiles/abWebsocket_peer.c]: For what do you need the WS_-defines in this file?

File: src/autobahnfiles/abWebsocket_peer.c [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+#define WS_CLOSE_FRAME 0x8
+#define WS_PING_FRAME 0x9
+#define WS_PONG_FRAME 0x0a
+
+uint8_t *binary_frame_buffer = NULL;
+size_t binary_frame_buffer_size = 0;
+char *text_frame_buffer = NULL;
+size_t text_frame_buffer_size = 0;
+struct cjet_utf8_checker checker;
+struct cjet_utf8_checker *checker_ptr;
```
Comment (gatzka) [inline src/autobahnfiles/abWebsocket_peer.c]: Why not just using &checker instead of checker_ptr?

File: src/autobahnfiles/abWebsocket_peer.c [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+#define CRLF "\r\n"
+
+#define WS_CONTINUATION_FRAME 0x0
+#define WS_TEXT_FRAME 0x1
+#define WS_BINARY_FRAME 0x2
+#define WS_CLOSE_FRAME 0x8
+#define WS_PING_FRAME 0x9
+#define WS_PONG_FRAME 0x0a
+
+uint8_t *binary_frame_buffer = NULL;
```
Comment (gatzka) [inline src/autobahnfiles/abWebsocket_peer.c]: I think all these globals can be made static to not spoiling the global namespace.

In addition, it is a bit ugly that is global at all. I think all theses variables should be members of the peer. To do this you have to write your own struct ab_ws_peer which contains all these variables.

File: src/tests/utf8_checker_test.cpp [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	{
+	}
+
+	struct cjet_utf8_checker c;
+};
+
+/**
+ * @brief tests the utf8 checker with a message of length 0
+ */
+BOOST_AUTO_TEST_CASE(test_message_zero_length)
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: Using BOOST_FIXTURE_TEST_CASE you don't have to write F f yourself and you have access to all F members automatically.

File: src/tests/utf8_checker_test.cpp [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	ret = cjet_is_byte_sequence_valid(&f.c, valid_message_long + sizeof(valid_message_long) - 1, 1, true);
+	BOOST_CHECK_MESSAGE(ret == true, "Message should be valid!");
+}
+
+/**
+ * @brief tests the utf8 checker with a invalid message
+ */
+BOOST_AUTO_TEST_CASE(test_invalid_message)
+{
+	int ret;
```
Comment (gatzka) [inline src/tests/utf8_checker_test.cpp]: ret could be declared inside the loop

File: src/tests/websocket_frame_test.cpp [hunk 10]
```diff
@@ -451,6 +495,149 @@ BOOST_AUTO_TEST_CASE(test_receive_binary_message)
...
+
+	prepare_message_string_frag(WS_OPCODE_CONTINUATION, messages[1], is_server, mask, false);
+	ws_get_header(&f.ws, read_buffer_ptr++, read_buffer_length);
+
+	BOOST_CHECK_MESSAGE(text_frame_received_called == 0, "Callback for text frames was called: " << text_frame_received_called);
+	BOOST_CHECK_MESSAGE(got_error, "Did not got an error when receiving a message without start frame!");
+	BOOST_CHECK_MESSAGE(is_close_frame(WS_CLOSE_PROTOCOL_ERROR), "No close frame sent after error!");
+	BOOST_CHECK_MESSAGE(br_close_called, "buffered_reader not closed after websocket close!");
+	BOOST_CHECK_MESSAGE(readback_buffer_length == 0, "Received a message!");
+
```
Comment (gatzka) [inline src/tests/websocket_frame_test.cpp]: Take out empty line.

File: src/tests/websocket_frame_test.cpp [hunk 11]
```diff
@@ -507,6 +713,29 @@ BOOST_AUTO_TEST_CASE(test_receive_pong_on_client)
...
+/*
+ * tests if the connection is failed immediately, when any rsv-bit is set.
+ * It doesn't matter if it is a text, binary or ping, because it is
+ * shut down before the type is determined.
+ */
+BOOST_AUTO_TEST_CASE(test_receive_text_message_with_rsv_bit_set)
+{
+	bool is_server = true;
+
+	char message [] = "Hello World!";
```
Comment (gatzka) [inline src/tests/websocket_frame_test.cpp]: Reduce scope of variable.

**Review by gatzka — CHANGES_REQUESTED** [review]
Overall good, some minor coding style stuff.
The only thing that needs real rework is to create an own struct ab_ws_peer to eliminate the global variables like buffers.

File: src/CMakeLists.txt (hunk 1 above)
Comment (felix-re) [reply src/CMakeLists.txt]: done

File: src/autobahnfiles/abMain.c (hunk 2 above)
Comment (felix-re) [reply src/autobahnfiles/abMain.c]: done

File: src/autobahnfiles/abMain.c (hunk 3 above)
Comment (felix-re) [reply src/autobahnfiles/abMain.c]: done

File: src/autobahnfiles/abMain.c (hunk 4 above)
Comment (felix-re) [reply src/autobahnfiles/abMain.c]: done

File: src/autobahnfiles/abWebsocket_peer.c (hunk 5 above)
Comment (felix-re) [reply src/autobahnfiles/abWebsocket_peer.c]: relics from websocket.c, they're not used anymore.
Completely removed.

File: src/autobahnfiles/abWebsocket_peer.c (hunk 6 above)
Comment (felix-re) [reply src/autobahnfiles/abWebsocket_peer.c]: done

File: src/autobahnfiles/abWebsocket_peer.c (hunk 7 above)
Comment (felix-re) [reply src/autobahnfiles/abWebsocket_peer.c]: done

File: src/tests/utf8_checker_test.cpp (hunk 8 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: done

File: src/tests/utf8_checker_test.cpp (hunk 9 above)
Comment (felix-re) [reply src/tests/utf8_checker_test.cpp]: done

File: src/tests/websocket_frame_test.cpp (hunk 10 above)
Comment (felix-re) [reply src/tests/websocket_frame_test.cpp]: done

File: src/tests/websocket_frame_test.cpp (hunk 11 above)
Comment (felix-re) [reply src/tests/websocket_frame_test.cpp]: Do you mean remove empty line?

File: src/autobahnfiles/abWebsocket_peer.h [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef CJET_WEBSOCKET_PEER_H
+#define CJET_WEBSOCKET_PEER_H
+
+#include <stdint.h>
+
```
Comment (gatzka) [inline src/autobahnfiles/abWebsocket_peer.h]: #include <stddef.h> is missing. It's necessary for size_t.

**Review by gatzka — CHANGES_REQUESTED** [review]
No further objections, please just add the missing include.

File: src/autobahnfiles/abWebsocket_peer.h (hunk 12 above)
Comment (felix-re) [reply src/autobahnfiles/abWebsocket_peer.h]: thanks, done

**Pull request merged**